src/contact.rs

```diff
@@ -466,12 +468,14 @@ impl Contact {
     /// Depending on the origin, both, "row_name" and "row_authname" are updated from "name".
     ///
     /// Returns the contact_id and a `Modifier` value indicating if a modification occurred.
+    ///
+    /// Returns None if the contact with such address cannot exist.
     pub(crate) async fn add_or_lookup(
         context: &Context,
         name: &str,
         addr: &str,
         mut origin: Origin,
-    ) -> Result<(ContactId, Modifier)> {
+    ) -> Result<Option<(ContactId, Modifier)>> {
```

Comment:
Is this approach better than defining a custom error type:
```
pub(crate) enum AddOrLookupError {
    InvalAddr,
    Other(anyhow::Error),
}
```
? I think the latter is more common, we can define additional "error codes" if we'll need. And i guess we already have other functions that require more than one "special" error. And we can also implement From<AddOrLookupError> for anyhow::Error to ease its handling

Reply:
> The problem with using enums for errors is that it is easy to forget that some error variants, like the proposed `InvalAddr` are actually not exceptional situations and should not be bubbled up. If you forget to do this and pass the error up with `?`, the connection is dropped and after reconnection we try to fetch and process the same message again in infinite loop.

We can avoid implementing `From<CustomError>` for `anyhow::Error` and add a custom conversion function instead so that an implicit conversion between error types with `?` is impossible and a user is forced to decide to either forward or just log it.

> You also cannot add a `.context` directly to the function result because it will convert to `anyhow` error type and have to match on the error type like this:
> 
> ```
> match add_or_lookup(...) {
>   Err(AddOrLookupError::InvalAddr) => ...,
>   Err(AddOrLookupError::Other(err)) => return Err(err).context(...),
>   Ok(contact_id) => ...
> }
> ```

I think a situation when a user gives us a bad addr is rather an error and needs contexting too. Particularly, if we forward it to the upper level where it is logged (but i don't know a good way to attach context to it, `InvalAddr(anyhow::Result)` looks strange). If there's no real need to forward contexted errors preserving their types, e.g. if we just log `InvalAddr` and forward other errors, there's no problem.

Reply:
Btw, personally i found contexting errors not very useful in my C experience. Better to context all logs and log all errors at the lowest possible level

Reply:
> I think it's impossible to delete or block existing generic implementation of `From` that is already in the `anyhow` library: https://docs.rs/anyhow/latest/anyhow/struct.Error.html#impl-From%3CE%3E-for-Error

Yes, but if our custom error isn't `StdError`, `?` will fail to convert it.

> Now that I think about it, probably a less hacky way to do validation is to only accept valid email addresses. So `add_or_lookup` should accept not a `String`, but something like a `struct ValidContactAddress(String)` that implements `FromStr` trait which errors out when the address is not valid. Relevant: https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ Going to try implementing this. In this case any error will be a real exception and it will be impossible to call the function with invalid input.

But anyway, this approach is better in this particular case, yes.